server/interceptor/interceptor.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package interceptor defines gRPC interceptors for Trillian.
+package interceptor
+
+import (
+	"strings"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/errors"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/trees"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// TreeInterceptor ensures that all requests pertaining a specific tree are of correct type and
+// respect the current tree state (frozen, deleted, etc).
+type TreeInterceptor struct {
+	Admin storage.AdminStorage
+}
+
+// UnaryInterceptor executes the TreeInterceptor logic for unary RPCs.
+func (i *TreeInterceptor) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	rpcInfo, err := getRPCInfo(req, info.FullMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	if rpcInfo.doesNotHaveTree {
+		return handler(ctx, req)
+	}
+
+	tree, err := trees.GetTree(ctx, i.Admin, rpcInfo.treeID, rpcInfo.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = trees.NewContext(ctx, tree)
+	return handler(ctx, req)
+}
+
+// rpcInfo contains information about an RPC, as extracted from its request message.
+type rpcInfo struct {
+	// doesNotHaveTree states whether the RPC is tied to a specific tree.
+	// Examples of RPCs without trees are CreateTree (no tree exists yet) and ListTrees
+	// (zero to many trees returned).
+	doesNotHaveTree bool
+	// treeID is the tree ID tied to this RPC, if any.
+	treeID int64
+	// opts is the trees.GetOpts appropriate to this RPC (TreeType, readonly vs readwrite, etc).
+	// opts is not set if doesNotHaveTree is true.
+	opts trees.GetOpts
+}
+
+// getRPCInfo returns the rpcInfo for the given request, or an error if the request is not mapped.
+// Full method is the full RPC method name, as acquired from grpc.UnaryServerInfo (e.g.,
+// /trillian.TrillianLog/GetInclusionProof).
```

Comment:
I'd say here is the proper place - it's not up to the service definition how RPCs get intercepted. Done.